### ISSUE TITLE
Non-static objects don't block navmesh

### ISSUE THREAD

**Issue by TenebreGaming** [TheGameCreators/GameGuruRepo#4898]
EXP build 2023.08.30:

I have a wall that I temporarily attached a behavior to to test something, so I set it to Physics On, Always Active, and Immoble. I cannot walk through it, but NPC characters can. Bringing up the AI Management shows that the navmesh isn't blocked (there should be an "edge" visible) and you can clearly see where the NPC is targetting me through the wall.

[upgrade_switch.zip](https://github.com/TheGameCreators/GameGuruRepo/files/12502467/upgrade_switch.zip)

**Comment by Necrym59** [comment]
Add a temporary namesh at that position when its dynamic

**Comment by AmenMoses** [comment]
i.e. use RDBlockNavMesh( x, y, z, radius, 1 ) to block and RDBlockNavMesh( x, y, z, radius, 0 ) to unblock.

For a large wall you should add multiple small radius blockers in a line otherwise you will either block too large an area or leave gaps.

You can use the physlib GetObjectDimensions function to get the size of the wall and utillib Rotate3D function to calculate offsets based upon rotation of the wall to calculate where to but the blockers.

If you do make a nice generic function then request that it be added to one of the libraries or alternatively you could create a new navlib.lua and put any other useful nav-mesh based functions you come up with in it and then ask for that to be included  as stock.

**Comment by TenebreGaming** [comment]
I had thought about doing the thing with RDBlockNavMesh( ) exactly how descibed. It's definitely doable. A new rectangular navmesh blocking function would be much better (and could also be used for the doors in place of the radial one).

**Comment by AmenMoses** [comment]
isimmobile doesn't actually mean that an object won't move btw, it means that the Bullet engine can't move it, i.e. it cannot be moved by physics interactions.

It can still be moved by a Lua script like the slide/rotate behaviour or whatever.

It also has the side effect of allowing for polygon collision, normally dynamic entities can only have simple collision shapes like box or sphere for example, but if isimmobile=yes hen the entity can be dynamic but also have polygon collision shape.  This should be avoided unless absolutely necessary though as it can kill performance.

**Comment by TenebreGaming** [comment]
That is true. Maybe there can be a new tick box added to add navmesh blocking? Or even a Lua function to add/remove the "footprint" of the object in the navmesh?

**Comment by AmenMoses** [comment]
Yes, the math would be a lot simpler if we could specify a blocking area by four points representing the footprint of the entity!

**Comment by AmenMoses** [comment]
This is a way more complex problem than I first appreciated.

For example if you have a three storey building and place a wall such that it blocks two storeys but not the third, how to calculate where to put the blockers is rather challenging to say the least.

I think I've sussed out the simple cases, i.e. unscaled entity simply sitting on a single nav-mesh plane though.

**Comment by TenebreGaming** [comment]
I wouldn't expect an entity to block the navmensh any other place than where it's rooted. Multilevel would be quite a bit more complex.

**Comment by LeeBamberTGC** [comment]
@TenebreGaming Thanks all for the thread conversation, and I can confirm that non-static objects are never meant to create obstructions in the calculated navmesh, even isimmobile objects as AmenMoses described can move and be removed.  Only static objects have the power to bake a navmesh into shape. A recent improvement to the navmesh system means blockers are not switching off polygons in the nav mesh (which was unpredictable and crude), but now defines a zone of HIGH COST, forcing any path around blockers to the point they give up on the path completely if there is no way around the high cost zone.  I have moved this to a feature request and added a new title as I agree a rectangular zone (rather than a circle) can be very useful for scenarios of controlling a path around a non static object :)

**Comment by TenebreGaming** [comment]
@LeeBamberTGC Thank you. This is an acceptable solution and is probably exactly what is needed to fix the blocking of closed doors. It will definitely help for large door like warehouse/garage doors where a radial blocker can cause issues.

**Comment by OobieDoob-Benoobie** [comment]
> Yes, the math would be a lot simpler if we could specify a blocking area by four points representing the footprint of the entity!

If this could be attached to the entity with a ticker box option, it would make it very user friendly.
Ive run into this but apply mesh collision to the object to remedy. (expensive)

**Comment by Necrym59** [comment]
I think the biggest problem is not letting the npc's see/sense through walls/doors static or non-static then they wouldnt make the attempt to start with, and the npc colision capsule should be used to stop passing a door (dynamic object) no need to block/unblock the navemesh as this is not a reliable solution without messing about with navmesh block shaping .

**Comment by LeeBamberTGC** [comment]
@TenebreGaming @OobieDoob-Benoobie  In next EXP build new V20 scripts for doors will allow the shape of the door to be used, and also its angle for better finer blocker descriptions:
![image](https://github.com/TheGameCreators/GameGuruRepo/assets/35840583/5e0ca8b6-8119-47ae-8881-1d294e919996)

**Comment by OobieDoob-Benoobie** [comment]
> @TenebreGaming @OobieDoob-Benoobie In next EXP build new V20 scripts for doors will allow the shape of the door to be used, and also its angle for better finer blocker descriptions:

I have a question:  If I have a wall with all texture settings at zero (invisible barrier) that will disappear via a trigger zone
elsewhere, will the nav mesh update to show the pathway clear once the wall is gone?

**Comment by LeeBamberTGC** [comment]
@OobieDoob-Benoobie It depends, what script and LUA commands are you associating with your pseudo wall object?

**Comment by OobieDoob-Benoobie** [comment]
@LeeBamberTGC If it has to be a door, I could just use Door slide and have the angle take it under the topography.

**Comment by TenebreGaming** [comment]
This seems to be working acceptably.
